feat: Gator permissions controller integration

V00D00-child opened the pull request:
## **Description**

This PR adds [@metamask/gator-permissions-controller](https://github.com/MetaMask/core/tree/main/packages/gator-permissions-controller) package into extension:

- `@metamask/gator-permissions-controller`: a controller to handle managing gator permissions in MetaMask clients with [@metamask/gator-permissions-snap](https://github.com/MetaMask/snap-7715-permissions/tree/main/packages/gator-permissions-snap) integration.

It also creates background actions to allow the UI to submit requests to the background to invoke actions exposed on `@metamask/gator-permissions-controller`:
- `fetchAndUpdateGatorPermissions`

### **Note - local snap for development**

The `GatorPermissionsControllerInit` is configured to use a local running instance of the gator-snap for non-production environments to support development in future PRs.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/35627?quickstart=1)

## **Related issues**

Requires: https://github.com/MetaMask/core/pull/6467
Relates to(MM Core): https://github.com/MetaMask/core/pull/6033
Relates to: #35307
Required by: #35219
## **Manual testing steps**

There is no need for testing, since we are integrating dependencies to support future PRs.

### **Before**

### **After**

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

V00D00-child committed:
Add gator permissions controller to extension

V00D00-child committed:
bugfix: GatorPermissionsController Initialization Error

V00D00-child committed:
Add GatorPermissionsController to CONTROLLER_MESSENGERS

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child committed:
Drop npm preview of gator-permissions-controller for initial published release

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child commented:
@metamaskbot update-policies

V00D00-child committed:
Drop manual policies update to allow metamaskbot to update policies

metamaskbot committed:
Update LavaMoat policies

V00D00-child committed:
Bugfix: Make GatorPermissionsController API Methods Available in all Builds

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child committed:
Resolve lint errors due to different BaseController versions

V00D00-child committed:
Fix lockfile dedupe

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child committed:
Add GatorPermissionsController to e2e state-snapshot

V00D00-child committed:
Add GatorPermissionsController state props to metamask e2e state

V00D00-child committed:
Fix typo in test

V00D00-child committed:
Fix lint errors and only include GatorPermissionsController in state-snapshot after json files

V00D00-child committed:
Use GATOR_PERMISSIONS_PROVIDER_SNAP_ID build var. Remove enableGatorPermissions and disableGatorPermissions from api

V00D00-child commented on app/scripts/controller-init/gator-permissions/gator-permissions-controller-init.ts:
```diff
@@ -0,0 +1,56 @@
+      enableGatorPermissions:
+        controller.enableGatorPermissions.bind(controller),
+      disableGatorPermissions:
+        controller.disableGatorPermissions.bind(controller),
```
These methods allow runtime mutation of the `isGatorPermissionsEnabled` in the controller state. 

These methods are no longer exposed, and we use the `GATOR_PERMISSIONS_ENABLED` flag to set `isGatorPermissionsEnabled` at build time, so runtime mutation is no longer necessary.

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child commented on app/scripts/controller-init/gator-permissions/gator-permissions-controller-init.ts:
```diff
@@ -0,0 +1,56 @@
+    return {
+      isGatorPermissionsEnabled: false,
+      isFetchingGatorPermissions: false,
+    };
```
Now using `GATOR_PERMISSIONS_PROVIDER_SNAP_ID` with a single codepath.

V00D00-child committed:
bugfix: Invalid SnapId Environment Variable Causes Errors

V00D00-child committed:
Ensure gatorPermissionsProviderSnapId is set undefined in e2e test

V00D00-child committed:
Remove gatorPermissionsProviderSnapId from e2e test to allow it to populate to undefined

V00D00-child committed:
Only valid snapId is set otherwise we rely on default value set in controller constructor

mathieuartu reviewed (dismissed):
Haven't tested, but code LGTM! 🙂

jeffsmale90 reviewed (commented):
One minor comment regarding the `snapId`, and one question:

What's the state lifecycle wrt `isGatorPermissionsEnabled` - is this generated every time the application starts?

If it were persisted, could it result in changes to the env variable not being reflected in the state?

V00D00-child committed:
Throws error when snapId is invalid

V00D00-child committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

V00D00-child committed:
Use empty string for GATOR_PERMISSIONS_PROVIDER_SNAP_ID

V00D00-child committed:
Remove throwing error

V00D00-child commented on app/scripts/controller-init/gator-permissions/gator-permissions-controller-init.ts:
```diff
@@ -2,28 +2,28 @@ import {
+    try {
+      assertIsValidSnapId(snapId);
+    } catch (error) {
+      return baseGatorPermissionsControllerState;
```
Not needed, the `GATOR_PERMISSIONS_PROVIDER_SNAP_ID` is empty/undefined in `builds.yml`. If we throw an error, then the controller will fail to initialize. The `GatorPermissionsController` will use the default if `gatorPermissionsProviderSnapId` is not set correctly by the client.

V00D00-child replied:
The controller is designed to fall back to a predefined `snapId` if misconfigured. This fail-safe is intentional to ensure robustness and prevent runtime issues.

> it is going to result in a build that's not configured as specified.

If the client provides an invalid configuration, the resulting build may not match it  because the configuration itself is incorrect. In this case, falling back to the default snapId is the expected behavior. It's not a misconfiguration of the build, but a safeguard against one.

jeffsmale90 reviewed (changes requested):
I've opened #35775 into this branch that rejects invalid snapId provided via configuration, but falls back to default if no configuration is provided. This ensures only valid configuration is accepted, but allows end-to-end tests to run without needing additional configuration to be supplied.

jeffsmale90 committed:
Reject invalid GATOR_PERMISSIONS_PROVIDER_SNAP_ID configuration (#35775)

## **Description**

When instantiating the `GatorPermissionsController`, previously invalid
snapIds specified by `process.env.GATOR_PERMISSIONS_PROVIDER_SNAP_ID`
would be ignored, falling back to the default value specified by the
controller implementation.

This is undesired because the configuration that is supplied at build
time is being ignored. The preferred behaviour is to fail, and require
valid configuration.

This change throws an error if an invalid snapId is specified, and adds
reasonable configuration values to builds.yml configuration. If no
values are specified, the default from Gator Permissions Controller is
used.

In order to enable the feature for any build (local, flask, production)
with the production snaps, only the `GATOR_PERMISSIONS_ENABLED` variable
would need to be set.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/35775?quickstart=1)

## **Changelog**

This PR is not user facing, as it's targeting a feature branch. Change
log details will be provided in that PR.

## **Related**

https://github.com/MetaMask/metamask-extension/pull/35627

## **Manual testing steps**

Provide explicit configuration (default):
1. Extension should start successfully

Provide no configuration:
1. Remove the `GATOR_PERMISSIONS_PROVIDER_SNAP_ID` from builds.yml
2. Extension should start successfully

Provide invalid configuration
1. Set `GATOR_PERMISSIONS_PROVIDER_SNAP_ID` to an invalid snapId (ie
"invalid-snap-id")
2. Extension should fail to start

jeffsmale90 committed:
Merge branch 'main' into feat/gator-permissions-controller-integration

jeffsmale90 reviewed (approved):
lgtm!
